application/startup.inc.php

```diff
@@ -68,9 +68,12 @@
 Session::Start();
 $oKPI->ComputeAndReport("Session Start");
 
-$sSwitchEnv = utils::ReadParam('switch_env', null);
+
+$sSwitchEnv = utils::ReadParam('switch_env', null, true);
 $bAllowCache = true;
-if (($sSwitchEnv != null) && file_exists(APPCONF.$sSwitchEnv.'/'.ITOP_CONFIG_FILE) &&( Session::Get('itop_env') !== $sSwitchEnv))
+if (($sSwitchEnv != null) && (file_exists(APPCONF.$sSwitchEnv.'/'.ITOP_CONFIG_FILE)) && 
+	(Session::Get('itop_env', null) !== $sSwitchEnv || Session::Get('itop_env', null) === null)
```

Comment:
If the session var is not set, the response will contain null and thus the first also already evaluate true.

```suggestion
	(Session::Get('itop_env', null) !== $sSwitchEnv)
```

Reply:
You mean the first part of the condition above?

`$sSwitchEnv = utils::ReadParam('switch_env', null, true);`

So if no switch_env is set, it would default to null.
If no itop_env is set as a session variable, it would default to null.
In this case, the condition ( null (itop_env) !== null (switch_env) ) would evaluate false, and the code would not be executed. 
Hence the second part of the condition: if no itop_env is set, forcefully do so.

But if I read the next part of the code, it would still be covered by the last "else"; but without resetting the opcache and APC.

I do think you're right, either I drop this part (since it's already covered at the end); or the last "else" doesn't really seem needed.

Reply:
If no switch is set, this if will also not be executed, see the first condition. My suggestion only removes redundant code.

Reply:
I see where it comes from. Tried to adjust it to the iTop 3 logic (original PR was for iTop 2.7), so this part is indeed redundant now since it's covered in iTop's newer session implementation.